src/testing/tokenizer.rs

```diff
@@ -107,17 +110,261 @@ impl Test {
 
         builders
     }
+
+    pub fn assert_valid(&self) {
+        for mut builder in self.builders() {
+            let mut tokenizer = builder.build();
+
+            // If there is no output, still do an (initial) next token so the parser can generate
+            // errors.
+            if self.output.is_empty() {
+                tokenizer.next_token();
+            }
+
+            // There can be multiple tokens to match. Make sure we match all of them
+            for expected_token in self.output.iter() {
+                let t = tokenizer.next_token();
+                self.assert_token(t, expected_token);
+            }
+
+            let borrowed_error_logger = tokenizer.error_logger.borrow();
+            assert_eq!(borrowed_error_logger.get_errors().len(), self.errors.len());
+
+            // Check error messages
+            for error in &self.errors {
+                self.assert_error(&tokenizer, error);
+            }
+        }
+    }
+
+    // Run through the parsing without making assertions, for use in benchmarking and in order to
+    // disclose any panics that might happen
+    pub fn tokenize(&self) {
+        for mut builder in self.builders() {
+            let mut tokenizer = builder.build();
+
+            for _ in self.output.iter() {
+                tokenizer.next_token();
+            }
+        }
+    }
+
+    fn assert_error(&self, tokenizer: &Tokenizer, expected: &Error) {
+        // Iterate all generated errors to see if we have an exact match
+        for actual in tokenizer.get_error_logger().get_errors() {
+            if actual.message == expected.code
+                && actual.line as i64 == expected.line
+                && actual.col as i64 == expected.col
+            {
+                return;
+            }
+        }
+
+        // Try and find an error that matches the code, but has a different line/pos. Even though
+        // it's not always correct, it might be a off-by-one position.
+        for actual in tokenizer.get_error_logger().get_errors() {
+            if actual.message == expected.code
+                && (actual.line as i64 != expected.line || actual.col as i64 != expected.col)
+            {
+                panic!(
+                    "[{}]: wanted {:?}, got {:?}",
+                    self.description, expected, actual
+                );
```

Comment:
It was convenient to move these `assert_` methods to `impl Test` in order to make it easier to print out additional context (`self.description`, in this case).